editor/src/core/shared/github.ts

```diff
@@ -348,3 +366,63 @@ export function githubFileChangesEquals(
     arrayEquals(a.deleted, b.deleted)
   )
 }
+
+export type GithubFileChangesListItem = {
+  status: GithubFileStatus
+  filename: string
+}
+
+export function githubFileChangesToList(
+  changes: GithubFileChanges | null,
+): Array<GithubFileChangesListItem> {
+  if (changes == null) {
+    return []
+  }
+
+  const toItem = (status: GithubFileStatus, files: Array<string>) =>
+    files.map((d) => ({ status: status, filename: d }))
+
+  const sortByFilename = (a: GithubFileChangesListItem, b: GithubFileChangesListItem) =>
+    a.filename.localeCompare(b.filename)
+
+  return [
+    ...toItem('untracked', changes.untracked),
+    ...toItem('modified', changes.modified),
+    ...toItem('deleted', changes.deleted),
+  ].sort(sortByFilename)
+}
+
+export function revertAllGithubFiles(
+  dispatch: EditorDispatch,
+  branchContents: ProjectContentTreeRoot | null,
+): void {
+  if (branchContents != null) {
+    dispatch([updateProjectContents(branchContents)], 'everyone')
+  }
+}
+
+export function revertGithubFile(
+  dispatch: EditorDispatch,
+  status: GithubFileStatus,
+  filename: string,
+  projectContents: ProjectContentTreeRoot,
+  branchContents: ProjectContentTreeRoot | null,
+): void {
+  if (branchContents != null) {
+    switch (status) {
+      case 'untracked':
+        dispatch([deleteFile(filename)], 'everyone')
+        break
+      case 'deleted':
+      case 'modified':
+        const newContents = treeToContents(projectContents)
+        const branchContentsTree = treeToContents(branchContents)
```

Comment:
Can use `getContentsTreeFileFromString` to retrieve a file rather than converting the structures and `addFileToProjectContents` to add it into another.

Reply:
https://github.com/concrete-utopia/utopia/pull/2743/commits/96b748b746c7283d34266309049dc424841ea20c